src/collections/Collections/Comparers/ArrayEqualityComparer.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace System.Collections.Generic;
+
+public sealed class ArrayEqualityComparer<T> : IEqualityComparer<T[]>
+{
+    private readonly IEqualityComparer<T>? comparer;
+
+    private IEqualityComparer<T> Comparer()
+        =>
+        comparer ?? EqualityComparer<T>.Default;
+
+    public ArrayEqualityComparer() { }
+
+    public ArrayEqualityComparer(IEqualityComparer<T>? comparer)
+        =>
+        this.comparer = comparer;
+
+    public static ArrayEqualityComparer<T> Default
+        =>
+        DefaultInstance.Value;
+
+    public bool Equals(T[]? x, T[]? y)
+    {
+        if (ReferenceEquals(x, y))
+        {
+            return true;
+        }
+
+        if (x is null || y is null)
+        {
+            return false;
+        }
+
+        if (x.Length != y.Length)
+        {
+            return false;
+        }
+
+        if (x.Length > 0)
+        {
+            var comparer = Comparer();
+
+            for (int i = 0; i < x.Length; i++)
+            {
+                if (comparer.Equals(x[i], y[i]) is false)
+                {
+                    return false;
+                }
+            }
+        }
+
+        return true;
+    }
+
+    public int GetHashCode([DisallowNull] T[] obj)
```

Comment:
DisallowNull?

Reply:
Actually yes)

`[DisallowNull]` is a part of contract of `IEqualityComparer<T>` since the nullable feature was brought into .NET Core+, and it is implemented automatically for every implementation of the interface.

Also, since .NET Framework 1.0, the comparers are expected to throw `ArgumentNullException` in `GetHashCode(T obj)` for null case in accordance to their declaration.

But actually the common best practice .NET for `GetHashCode(T obj)` is:
- To return zero for null on the comparer side (despite of `IEqualityComparer<T>` declaration);
- Not to pass null into `GetHashCode(T obj)` from a client side (the client side should handle the null case by itself).

So, it looks like a double check to prevent exceptions)

Also please see [referenceequalitycomparer.gethashcode](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.referenceequalitycomparer.gethashcode) and [runtimehelpers.gethashcod](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.gethashcode)

So, actually we have to keep `[DisallowNull]` to accord the `IEqualityComparer<T>` declaration, and at the same time we have to return zero result for the null case in `GetHashCode(T obj)`.